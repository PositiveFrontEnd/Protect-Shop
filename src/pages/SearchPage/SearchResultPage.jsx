import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { actionInputText } from "../../store/homeSlice";
import {
    actionGetOneProduct,
    actionGetThreeProducts,
    actionSearchProducts
} from "../../store/productsSlice";
import Button from "src/components/Button/Button";
import {
    selectorInputText,
    selectorSearchProducts,
    selectorThreeProducts,
    selectorToken
} from "../../store/selectors";
import { actionFavoriteForAll } from "../../store/favoriteSlice"
import PrimaryCard from "../../components/Main/Cards/PrimaryCard/PrimaryCard/PrimaryCard";
import SearchSort from "src/pages/SearchPage/SearchSort";
import { useNavigate } from "react-router";
import { Link } from "react-router-dom";
import ScrollUp from "src/components/Footer/Svg/scrollUp.svg?react";


const SearchResultPage = () => {

    const inputText = useSelector(selectorInputText);
    const token = useSelector(selectorToken);
    const dispatch = useDispatch();
    const handleInputChange = (event) => {
        dispatch(actionInputText(event.target.value));
    }

    const handleSearch = () => {
        dispatch(actionSearchProducts({ query: `${inputText}` }))
        navigate(`/search?request=${encodeURIComponent(inputText)}`)
        handleReset()
    }

    const [sort, setSort] = useState('');
    const handleSort = (event) => {
        setSort(event.target.value);
    };

    const products = useSelector(selectorSearchProducts);
    const sortedProducts = [...products];

    sortedProducts.sort((a, b) => {
        if (sort === 'lowToHigh') {
            if (a.currentPrice < b.currentPrice) {
                return -1
            }
            if (a.currentPrice > b.currentPrice) {
                return 1
            }
            return 0
        }
        if (sort === 'highToLow') {
            if (a.currentPrice > b.currentPrice) {
                return -1
            }
            if (a.currentPrice < b.currentPrice) {
                return 1
            }
            return 0
        }
        if (sort === '') {
            if (a.date > b.date) {
                return -1
            }
            if (a.date < b.date) {
                return 1
            }
            return 0
        }
    })

    const firstSortedProducts = sortedProducts.slice(0, 12);
    const secondSortedProducts = sortedProducts.slice(12, 24);
    const restSortedProducts = sortedProducts.slice(12);

    const handleEnterPress = (event) => {
        if (event.key === 'Enter') {
            return handleSearch()
        }
    }

    const handleSubmit = (e) => {
        e.preventDefault()
        handleReset()
    };

    const [clickCount, setClickCount] = useState(0);

    const handleIncrement = () => {
        setClickCount(clickCount + 1);
    };

    const handleReset = () => {
        setClickCount(0);
    };

    const navigate = useNavigate();
    const threeProducts = useSelector(selectorThreeProducts);
    const handleFavorite = (productId, event) => {
        event.stopPropagation();
        dispatch(actionFavoriteForAll({ productId, token }));
    };

    const handleProduct = (item) => {
        dispatch(actionGetOneProduct(item._id));
        dispatch(actionGetThreeProducts(item.name));
        navigate(
            `/catalogue/${item.categories}/${item.type}/${item._id}`
        );
    };
    const [isVisible, setIsVisible] = useState(false);
    useEffect(() => {
        window.addEventListener('scroll', toggleVisibility);
        return () => {
            window.removeEventListener('scroll', toggleVisibility);
        };
    }, []);
    const toggleVisibility = () => {
        if (window.pageYOffset > 700) {
            setIsVisible(true);
        } else {
            setIsVisible(false);
        }
    };
    const scrollToTop = () => {
        window.scrollTo({
            top: 0,
            behavior: 'smooth',
        });
    };

    return (
        <div className='container'>
            <h2 className='search__title'>search<span>{inputText}</span></h2>
            <div className='search__block'>
                <div className='search__input-group'>
                    <form onSubmit={handleSubmit}>
                        <input value={inputText} type='text' name='search'
                            onChange={handleInputChange}
                            className='search__input'
                            onKeyDown={handleEnterPress} />
                    </form>
                    <Button black onClick={handleSearch} className='search__btn'>SEARCH</Button>
                </div>
                <div className='search__function-group'>
                    <SearchSort handleSort={handleSort} sort={sort} />
                </div>
            </div>
            {sortedProducts.length === 0 ?
                <h3 className='search__result__title'>We donâ€™t have this goods
                    now<br /> Change your search request<br /> or look other similar goods in our <Link
                        to='/catalogue'><span>catalogue</span></Link></h3> :
                <h3 className='search__result__title'>We found this goods for you</h3>
            }
            <div className='search__result'>
                {firstSortedProducts.map((item) => (
                    <PrimaryCard key={item._id} card={item} arr={threeProducts}
                        handleProduct={handleProduct}
                        handleFavorite={handleFavorite}
                    />))}
                {clickCount === 1 && (
                    secondSortedProducts.map((item) => (
                        <PrimaryCard key={item._id} card={item} arr={threeProducts}
                            handleProduct={handleProduct}
                            handleFavorite={handleFavorite}
                        />)))}
                {clickCount === 2 && (
                    restSortedProducts.map((item) => (
                        <PrimaryCard key={item._id} card={item} arr={threeProducts}
                            handleProduct={handleProduct}
                            handleFavorite={handleFavorite}
                        />)))}
            </div>
            {sortedProducts.length > 12 && (
                <>
                    <div
                        className={`scroll__to__top ${isVisible ? 'visible' : ''}`}
                        onClick={scrollToTop}>
                        <ScrollUp />
                    </div>
                    <Button
                        className={clickCount === 2 || (clickCount === 1 && sortedProducts.length < 24) ? 'button-inactive' : 'button-white button__SeeMore-white'}
                        click={handleIncrement}>See more</Button>
                </>
            )}
        </div>
    );
};

export default SearchResultPage;