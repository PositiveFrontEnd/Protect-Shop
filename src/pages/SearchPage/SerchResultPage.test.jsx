import React from "react";
import { render } from "@testing-library/react";
import { Provider } from 'react-redux';
import SearchResultPage from "./SearchResultPage";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";

describe("Тестування компонента  SearchResultPage", () => {
    test("Знімок", () => {
        const mockStore = configureStore();
        const initialState = {
            home: {
                isAnimation: false,
                inputText: "some text",
                filterHomePage: [],
            },
            user: {
                token: "token",
                registrationModal: false,
                authorizationStatus: null,
                registrationStatus: "some text",
                registrationData: {},
                correctData: "some text",
                isAdmin: false,
            },
            products: {
                products: [],
                productsByCategory: [],
                productsByType: [],
                oneProduct: {},
                ThreeProducts: [],
                ThreeColors: [],
                filters: [],
                twelveProducts: [],
                filtersParameters: {},
                twelveFiltersProducts: [],
                searchProducts: [],
                youSee: [],
                previewProductInfo: {},
            },
            order: {
                orderState: [],
                productsForOrderGuest: {},
                infoForOrderGuest: [],
                deliveryForOrderGuest: {},
                order: {
                    letterSubject: "Thank you for order!"
                },
                promoCodePrice: "",
                orderGuest: { letterSubject: "Thank you for order!" },
            },
            basket: {
                basket: [],
                priceGuest: "''",
                guestBasket: {},
            },
            catalogue: {
                sortCatalogue: "some text",
                loadFilter: false,
            },
            favorite: {
                favorite: [],
                favoriteForCustomer: [],
                guestFavorite: [],
            },
            message: {
                letters: [],
            }

        };

        const store = mockStore(initialState);
        const searchResultPage = render(
            <Provider store={store}>
                <BrowserRouter>
                    <SearchResultPage />
                </BrowserRouter>
            </Provider>
        );

        expect(searchResultPage).toMatchSnapshot();
    });
});
