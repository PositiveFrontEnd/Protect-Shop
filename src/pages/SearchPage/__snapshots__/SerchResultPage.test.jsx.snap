// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Тестування компонента  SearchResultPage > Знімок 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <h2
          class="search__title"
        >
          search
          <span>
            some text
          </span>
        </h2>
        <div
          class="search__block"
        >
          <div
            class="search__input-group"
          >
            <form>
              <input
                class="search__input"
                name="search"
                type="text"
                value="some text"
              />
            </form>
            <button
              class="button-black search__btn"
              type="button"
            >
              SEARCH
            </button>
          </div>
          <div
            class="search__function-group"
          >
            <div>
              <button
                class="sort"
              >
                sort
                <svg
                  fill="none"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.13828 6.54375L5.10391 5.57812L8.00078 8.475L10.8977 5.57812L11.8633 6.54375L8.00078 10.4062L4.13828 6.54375Z"
                    fill="#191919"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <h3
          class="search__result__title"
        >
          We don’t have this goods now
          <br />
           Change your search request
          <br />
           or look other similar goods in our 
          <a
            href="/catalogue"
          >
            <span>
              catalogue
            </span>
          </a>
        </h3>
        <div
          class="search__result"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <h2
        class="search__title"
      >
        search
        <span>
          some text
        </span>
      </h2>
      <div
        class="search__block"
      >
        <div
          class="search__input-group"
        >
          <form>
            <input
              class="search__input"
              name="search"
              type="text"
              value="some text"
            />
          </form>
          <button
            class="button-black search__btn"
            type="button"
          >
            SEARCH
          </button>
        </div>
        <div
          class="search__function-group"
        >
          <div>
            <button
              class="sort"
            >
              sort
              <svg
                fill="none"
                height="16"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4.13828 6.54375L5.10391 5.57812L8.00078 8.475L10.8977 5.57812L11.8633 6.54375L8.00078 10.4062L4.13828 6.54375Z"
                  fill="#191919"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <h3
        class="search__result__title"
      >
        We don’t have this goods now
        <br />
         Change your search request
        <br />
         or look other similar goods in our 
        <a
          href="/catalogue"
        >
          <span>
            catalogue
          </span>
        </a>
      </h3>
      <div
        class="search__result"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
